services/identity/UsersService.ts

```diff
@@ -1,56 +1,83 @@
-const { sequelize } = require("@database/models")
+import { Attributes } from "sequelize"
+
+import sequelize from "@database/index"
+import { User } from "@database/models"
+
+import MailClient from "./MailClient"
+import SmsClient from "./SmsClient"
+import TotpGenerator from "./TotpGenerator"
 
 // Allowed domains is a semicolon separate list of domains (e.g. .gov.sg, @agency.com.sg, etc)
 // that are allowed to login.
 const { DOMAIN_WHITELIST } = process.env
 
+interface UsersServiceProps {
+  otp: TotpGenerator
+  mailer: MailClient
+  smsClient: SmsClient
+  repository: Attributes<User>
+}
+
 class UsersService {
-  constructor({ otp, mailer, smsClient, repository }) {
-    this.repository = repository
+  // NOTE: Explicitly specifying using keyed properties to ensure
+  // that the types are synced.
+  private readonly otp: UsersServiceProps["otp"]
+
+  private readonly mailer: UsersServiceProps["mailer"]
+
+  private readonly smsClient: UsersServiceProps["smsClient"]
+
+  private readonly repository: UsersServiceProps["repository"]
+
+  private readonly whitelistDomains: string[]
+
+  constructor({ otp, mailer, smsClient, repository }: UsersServiceProps) {
     this.otp = otp
     this.mailer = mailer
     this.smsClient = smsClient
+    this.repository = repository
 
     this.whitelistDomains = (DOMAIN_WHITELIST || ".gov.sg")
       .split(";")
       .map((domain) => domain.toLowerCase().trim())
   }
 
-  async findByEmail(email) {
+  async findByEmail(email: string) {
     return this.repository.findOne({ where: { email } })
   }
 
-  async findByGitHubId(githubId) {
+  async findByGitHubId(githubId: string) {
     return this.repository.findOne({ where: { githubId } })
   }
 
-  async updateUserByGitHubId(githubId, user) {
+  async updateUserByGitHubId(githubId: string, user: string) {
     await this.repository.update(user, { where: { githubId } })
   }
 
-  async findOrCreate(githubId) {
+  async findOrCreate(githubId: string | undefined) {
     const [user] = await this.repository.findOrCreate({
       where: { githubId },
     })
     return user
   }
 
-  async login(githubId) {
+  async login(githubId: string) {
     return sequelize.transaction(async (transaction) => {
-      const user = await this.findOrCreate(githubId, { transaction })
+      // NOTE: The service's findOrCreate is not being used here as this requires an explicit transaction
+      const user = await this.repository.findOrCreate(githubId, { transaction })
```

Comment:
Could we consider removing the `this.findOrCreate()` method since we don't use it anywhere else? It could get quite confusing.

Also it looks like we are not using the transaction in `this.findOrCreate()`.

The new login method could look something like this

```
 async login(githubId: string) {
    return sequelize.transaction(async (transaction) => {
      const [user] = await this.repository.findOrCreate({
        where: { githubId },
        transaction
      })
      user.lastLoggedIn = new Date()
      return user.save({ transaction })
   }
}
```

Reply:
Ah I see this was resolved in https://github.com/isomerpages/isomercms-backend/pull/378/commits/593355d275ddbba72f76e198faf6905d03ce093d

Reply:
But we could still remove this.findOrCreate() because it isn't used anywhere else

Reply:
i'm not opposed to removing it but this does mean that users will only ever be created through logging in/manually. this could potentially mean that if we wish to allow agency users to create and manage their own users, it would have to be added back in. 

wdyt about this scenario or should we just remove if first?

Reply:
Hmm that's a good question. This might be a question for @kathleenkhy and @NatMaeTan.

Do we allow admin users to create other users in our system? In GitHub for example, an admin user cannot add a new user if that new user hasn't created their GitHub account. My first thought is that we might want to follow GitHub's model because we might have gibberish users in the system (i.e. there is no way to validate that gibberish@tech.gov.sg is not a real human).

Given that we aren't entirely sure what to do just yet, I'm in favor of removing this method for now. Thoughts?

Reply:
I'm thinking we may not need to have such a control right now? Have checked with @jacksonOGP that FormSG doesn't check on the validity of emails when adding new collaborators (as long as the email domain ends with gov.sg or a whitelisted domain) and agencies aren't really bothered with it as well.

If our concern is on security or dirty data that gibberish emails are added, : 
1. Security: at the end of the day as long as the email belongs to a valid officer that can log in with the verified OTP, I don't think there should be much of a risk in terms of security (or I don't see a motive for a valid officer to deface/ harm another agency user's site)
2. Dirty data: Agency users should keep their list of team members with access to repos clean and should remove invalid users that have been granted access to the repos (gibberish emails included). We could emphasise on this further after we launch Identity as well but if I'm not wrong, this is an IM8 requirement (though we do have a waiver, agencies should abide to this).

Let me know if this makes sense or if you might have other views as well? @seaerchin @prestonlimlianjie

Reply:
Will take a look at FormSG's implementation of this feature and get back to you